lighthouse-core/audits/time-to-interactive.js

```diff
@@ -0,0 +1,164 @@
+   *     AKA: First meaningful paint has fired.
+   *   - Page is nearly visually complete
+   *     Visual completion is 85%
+   *   - User-agent loading indicator is done
+   *     Current definition: Top frame and all iframes have fired window load event
+   *     Proposed definition (from cl/1860743002): top frame only: DCL ended and all layout-blocking resources, plus images that begun their request before DCL ended have finished.
+   *     Alternative definition (from Chrome on Android Progress Bar Enhancements - google-only, sry): top frame's DOMContentLoaded + top frame's images (who started before DCL) are loaded
+   *
+   * The page is actually ready for user:
+   *   - domContentLoadedEventEnd has fired
+   *     Definition: HTML parsing has finished, all DOMContentLoaded handlers have run.
+   *     No risk of DCL event handlers changing the page
+   *     No surprises of inactive buttons/actions as DOM element event handlers should be bound
+   *   - The main thread is available enough to handle user input
+   *     first 500ms window where Est Input Latency is <50ms at the 90% percentile.
+   * @param {!Artifacts} artifacts The artifacts from the gather phase.
+   * @return {!AuditResult} The score from the audit, ranging from 0-100.
+   */
+  static audit(artifacts) {
+    // We start looking at Math.Max(FMPMetric, visProgress[0.85])
+    return FMPMetric.audit(artifacts).then(fmpResult => {
+      if (fmpResult.value === -1) {
+        return generateError(fmpResult.debugString);
+      }
+      const fmpTiming = parseFloat(fmpResult.rawValue);
+      const timings = fmpResult.extendedInfo.timings;
+
+      // Process the trace
+      const tracingProcessor = new TracingProcessor();
+      const model = tracingProcessor.init(artifacts.traceContents);
+      const endOfTraceTime = model.bounds.max;
+
+      // TODO: Wait for DOMContentLoadedEndEvent
+      // TODO: Wait for UA loading indicator to be done
+
+      // look at speedline results for 85% starting at FMP
+      const visualProgress = artifacts.Speedline.frames.map(frame => {
+        return {
+          progress: frame.getProgress(),
+          time: frame.getTimeStamp()
+        };
+      });
+      const fMPts = timings.fMPfull + timings.navStart;
+      const visuallyReady = visualProgress.find(frame => {
+        return frame.time >= fMPts && frame.progress >= 85;
+      });
+      const visuallyReadyTiming = visuallyReady.time - timings.navStart;
+
+      // Find first 500ms window where Est Input Latency is <50ms at the 90% percentile.
+      let startTime = Math.max(fmpTiming, visuallyReadyTiming) - 50;
+      let endTime;
+      let currentLatency = Infinity;
+      const percentile = 0.9;
+      const threshold = 50;
+      let foundLatencies = [];
+
+      // When we've found a latency that's good enough, we're good.
+      while (currentLatency > threshold) {
+        // While latency is too high, increment just 50ms and look again.
+        startTime += 50;
```

Comment:
FWIW, `getRiskToResponsiveness` is very fast, and even for the very worst sites and even if somehow there was no windows with 90% at less than 50ms (so you had to go through the whole trace) you'd still be looking at a few hundred ms max for this loop if you dropped the increment to, say, 25ms at a time.

We could write a special iterative `getRiskToResponsiveness` just for calculating TTI and get that even faster, but it sounded like you wanted to avoid doing that :)